### ISSUE TITLE
COLMAP reconstructions producing incomplete/no results

### ISSUE THREAD

**Issue by Joeljensen28** [colmap/colmap#2866]
### Discussed in https://github.com/colmap/colmap/discussions/2850

<div type='discussions-op-text'>

<sup>Originally posted by **Joeljensen28** October 22, 2024</sup>
I'm working on a project where I am trying to build 3D models of rooms within homes, but I'm running into a few issues.

When I run `automatic_reconstructor`, the results are often incomplete, despite knowing that I have more data than what is present. See below:

<img width="552" alt="Screenshot 2024-10-22 at 2 53 24 PM" src="https://github.com/user-attachments/assets/87470e74-0d6c-46a9-a174-f0f2b2885912">

This mesh had about 600 pictures.

I also tried running a reconstruction process using known camera poses, and the results were negligible (all mesh outputs were <1kb and nothing was visible when I viewed it in MeshLab). Here are the detailed steps of what I did to make the mesh:

1. Create a workspace with the following heirarchy:
    ```
    workspace
    --dense
    --triangulated
    --images
    ----{image dataset here}
    --sparse
    ----cameras.txt
    ----images.txt
    ----points3D.txt
    ```
(I got the camera ex/intrinsics using NeRF capture mobile app)

2. Run these commands:
    ```
    cd workspace

    colmap feature_extractor \
        --database_path database.db \
        --image_path images

    colmap exhaustive_matcher --database_path database.db

    colmap point_triangulator --database_path database.db --image_path images --input_path sparse --output_path triangulated

    colmap image_undistorter \
        --image_path images \
        --input_path triangulated \
        --output_path dense

    colmap patch_match_stereo \
        --workspace_path dense

    colmap stereo_fusion \
        --workspace_path dense \
        --output_path dense/fused.ply

    colmap poisson_mesher \
        --input_path dense/fused.ply \
        --output_path dense/meshed.ply
    ```

Is COLMAP just not good at reconstructing 3D spaces in this fashion? I get pretty solid results when I try to make a reconstruction of a single object (like a water bottle), but I have not had any luck with 3D spaces. Is there something I'm doing wrong?</div>

**Comment by Joeljensen28** [comment]
For reference, here is what my `images.txt` looks like:

```
1 0.6544702804558814 -0.2422991669530667 -0.25398862020779817 -0.6696636777680918 0.004049856 0.0035372672 0.017289646 1 0.png

2 0.6551008660885742 -0.2683022773263062 -0.22958180989439414 -0.6679437551542443 0.02639354 0.018115241 0.008854851 1 1.png

3 0.6685229360313676 -0.3364716494774769 -0.17368574620535068 -0.6400758701567157 0.09006825 0.027616521 0.0041084215 1 2.png

4 0.6757010822101738 -0.35854234790857387 -0.12398870181761945 -0.6320617744210704 0.13989438 0.04109435 -0.0073726624 1 3.png

5 0.6793876345651281 -0.36596865669942535 -0.1229473694402259 -0.6240057861391054 0.1303029 0.036458652 -0.017646383 1 4.png

...
```

And `cameras.txt`:

`1 SIMPLE_PINHOLE 300 400 245.221 145.214 199.071`

Note that I am compressing the images for faster processing time, and thereby updating the camera intrinsics manually. The original intrinsics were `1 SIMPLE_PINHOLE 1920 1440 1582.8649 935.48224 716.72644`. Not sure if this affects the output.

**Comment by ahojnnes** [comment]
Some comments:

1. COLMAP should produce good results for well-textured scenes.
2. Point triangulator will keep the camera poses fixed. I assume the camera poses from the NeRF capture app extract camera poses from ARKit/ARCore. These poses are typically not accurate enough and will benefit from at least a refinement. You can insert a bundle adjustment step after the triangulation and/or interleave multiple rounds of (re-)triangulation and bundle adjustment depending on how good or bad the given poses are. If the initial poses are too far off, this approach may still not converge and you will be better off running the mapper from scratch.
3. An image resolution of 300x400 is generally too low resolution for colmap sparse and dense reconstruction. I would not recommend going below VGA resolution. If you can afford, keep the resolution at least at HD.

**Comment by Joeljensen28** [comment]
> 3\. An image resolution of 300x400 is generally too low resolution for colmap sparse and dense reconstruction. I would not recommend going below VGA resolution. If you can afford, keep the resolution at least at HD.

That is super good to know. Since upping the resolution, I have been getting much better results.

Do you know how I can deal with missing textures? I.e. the door/wall in this picture?
![image](https://github.com/user-attachments/assets/322d2f89-bbfa-4d96-9181-c22615e78631)

**Comment by hyqshr** [comment]
@Joeljensen28 Did you find a solution for missing walls/textures?

**Comment by Joeljensen28** [comment]
@hyqshr Not quite, I just closed this because my main question was answered. But I have a few things I'm wanting to try. I've been looking into using monocular depth estimation to get external depth maps for my photos, which when used on a single photo at a time outside of a traditional SfM pipeline produces [miraculous results](https://github.com/user-attachments/assets/b8fce58d-8878-4314-b0de-c07ef6a2aa92).

My current quest is trying to figure out a way to implement these external depth maps into an SfM pipeline to speed up the process and improve results, which is why I opened #2996. @ahojnnes How difficult would it be to implement something like this into COLMAP? As mentioned in my other issue, I'm a pretty junior programmer, but I would be excited to help out with implementation in any way I can.
